doc/manual/methods.rst

```diff
@@ -440,6 +440,212 @@ can also constrain type parameters of methods::
+method selection that identifies classes of types.
+We could do this by writing out a ``Union`` of the types with similar characteristics.
+But then this list would not be extensible.
+However, by expressing the property as a "trait",
+it is possible to express type behaviors in the abstract
+in a way that is flexible to new additions, but which has no performance impact.
+
+A trait is implemented by defining a pure generic function which returns a
+singleton value from a particular set, as a computation on the types of its arguments.
+
+The example above glossed over the implementation details of ``map`` and ``promote``,
+since those both operate in terms of type traits.
+When iterating over a matrix, such as in the implementation of ``map``,
+one important question is what order to use to traverse the data.
+When ``AbstractArray`` subtypes implement the ``linearindexing`` trait,
+other functions such as ``map`` can dispatch on this information to pick the best algorithm.
+This means that each subtypes does not need to implement a custom version of ``map``,
+since the generic definitions + trait classes will enable the system to select the fastest version.
+
+::
+
+    map(a::AbstractArray, b::AbstractArray) = map(Base.linearindexing(a, b), a, b)
+    map(::LinearSlow, a::AbstractArray, b::AbstractArray) = # generic implementation
+    map(::LinearFast, a::AbstractArray, b::AbstractArray) = # linear-fast implementation
+
+This trait-based mechanism is also present in the ``promote`` mechanism used by ``+``.
+That mechanism uses the types to compute the optimal common type for computing the operation.
+This makes it possible to reduce the problem of implementing every function for every pair of possible type arguments,
+to the much smaller problem of implementing a conversion operation from each type to a common type,
+plus a table of preferred pair-wise promotion rules.
+
+
+5. Output-type Computation
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The discussion of trait-based promotion provides a transition into our next design pattern:
+computing the output element type for a matrix operation.
+
+For implementing primitive operations, such as addition,
+we use the ``promote_type`` function to compute the desired output type.
+(As before, we saw this at work in the ``promote`` call in the call to ``+``).
+
+For more complex functions on matrices, it may be necessary to compute the expected return
+type for a more complex sequence of operations.
+This is often performed by the following steps:
+
+1. Write a small function ``op`` that expresses the set of operations performed by the kernel of the algorithm.
+2. Compute the element type ``R`` of the result matrix as ``promote_op(op, argument_types...)``,
+   where ``argument_types`` is computed from ``eltype`` applied to each input array.
+3. Build the output matrix as ``similar(R, dims)``, where ``dims`` is the desired dimensions of the output array.
```

Comment:
`similar(R, dims)` does not correspond to any `similar` signature?

Re. "`dims` is the desired dimensions", should that be "`dims` are the desired dimensions"?